File: 16_Intervals/03_Non_Overlapping_Intervals/0435-non-overlapping-intervals.py
"""
Problem: LeetCode 435 - Non-overlapping Intervals
 
Key Idea:
We can solve this problem by first sorting the intervals based on their end times. Then, we iterate through the sorted intervals and count the number of overlapping intervals. If the start of the current interval is less than the end of the previous interval, it means there is an overlap, so we increment the count and skip adding the current interval to the non-overlapping set. If there is no overlap, we update the end of the previous interval to be the end of the current interval.

Time Complexity:
- The time complexity of this approach is O(n log n) due to the sorting step, where n is the number of intervals.

Space Complexity:
- The space complexity is O(1), as we use only a constant amount of extra space.
"""


class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0

        intervals.sort(key=lambda x: x[1])
        non_overlapping = 1  # Count of non-overlapping intervals
        prev_end = intervals[0][1]

        for i in range(1, len(intervals)):
            if intervals[i][0] >= prev_end:
                non_overlapping += 1
                prev_end = intervals[i][1]

        return len(intervals) - non_overlapping

File: 16_Intervals/03_Non_Overlapping_Intervals/0435-non-overlapping-intervals.rs
/*
Problem: LeetCode 435 - Non-overlapping Intervals

Key Idea:
The key idea is to sort the intervals by their end times and then use greedy selection. Choose intervals that end earliest and remove overlapping intervals. This ensures that you retain as many non-overlapping intervals as possible.

Approach:
1. Sort the intervals based on their end values in ascending order.
2. Initialize a variable 'count' to 0 to keep track of non-overlapping intervals.
3. Initialize a variable 'prev_end' to -inf to store the end value of the last added interval.
4. Iterate through the sorted intervals:
   - If the current interval's start is greater than or equal to 'prev_end', it means it does not overlap with the previous interval.
     - Increment 'count' by 1.
     - Update 'prev_end' to the end value of the current interval.
   - If the current interval overlaps with the previous interval, skip it.
5. Return the value of 'count' as the maximum number of non-overlapping intervals.

Time Complexity:
O(n log n), where n is the number of intervals. The sorting step takes O(n log n) time.

Space Complexity:
O(1), as we use only a constant amount of extra space.
*/

impl Solution {
    pub fn erase_overlap_intervals(intervals: Vec<Vec<i32>>) -> i32 {
        if intervals.is_empty() {
            return 0;
        }

        let mut intervals = intervals;
        intervals.sort_by_key(|x| x[1]);

        let mut count = 1;
        let mut prev_end = intervals[0][1];

        for i in 1..intervals.len() {
            if intervals[i][0] >= prev_end {
                count += 1;
                prev_end = intervals[i][1];
            }
        }

        intervals.len() as i32 - count
    }
}
